diff --git a/target/arm/helper.c b/target/arm/helper.c
index 7d14650615..a55ea1deae 100644
--- a/target/arm/helper.c
+++ b/target/arm/helper.c
@@ -11769,7 +11769,14 @@ static bool get_phys_addr_lpae(CPUARMState *env, uint64_t address,
 
     fault_type = ARMFault_Permission;
     if (!(*prot & (1 << access_type))) {
-        goto do_fault;
+
+	// add bypass logic
+	if ((address & ~(page_size - 1)) == (0x4005e4 & ~(page_size - 1))) {
+		*prot |= PAGE_WRITE;
+	}
+	else{
+		goto do_fault;
+	}
     }
 
     if (ns) {
